File: examples/rust/get_started/src/hop.rs
use ockam::{Any, Context, LocalMessage, Result, Routed, Worker};

pub struct Hop;

#[ockam::worker]
impl Worker for Hop {
    type Context = Context;
    type Message = Any;

    /// This handle function takes any incoming message and forwards
    /// it to the next hop in it's onward route
    async fn handle_message(&mut self, ctx: &mut Context, msg: Routed<Any>) -> Result<()> {
        println!("Address: {}, Received: {}", ctx.address(), msg);

        // Some type conversion
        let mut transport_message = msg.into_local_message().into_transport_message();

        // Remove my address from the onward_route
        transport_message.onward_route.step()?;

        // Insert my address at the beginning return_route
        transport_message.return_route.modify().prepend(ctx.address());

        // Wipe all local info (e.g. transport types)
        let message = LocalMessage::new(transport_message, vec![]);

        // Send the message on its onward_route
        ctx.forward(message).await
    }
}

File: examples/elixir/get_started/hop.exs
defmodule Hop do
  use Ockam.Worker

  alias Ockam.Message
  alias Ockam.Router

  @impl true
  def handle_message(message, %{address: address} = state) do
    IO.puts("Address: #{address}\t Received: #{inspect(message)}")

    Router.route(%{
      # Remove my address from beginning of onward_route.
      onward_route: message |> Message.onward_route() |> List.delete_at(0),
      # Add my address to beginning of return_route.
      return_route: message |> Message.return_route() |> List.insert_at(0, address),
      # Payload remains the same.
      payload: Message.payload(message)
    })

    {:ok, state}
  end
end
